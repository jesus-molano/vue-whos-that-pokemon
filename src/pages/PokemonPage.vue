<script>
import PokemonImage from '@/components/PokemonImage'
import PokemonOptions from '@/components/PokemonOptions'
import { getPokemonOptions } from '@/helpers/getPokemonOptions'

export default {
  components: {
    PokemonImage,
    PokemonOptions
  },
  data () {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswers: false,
      message: null
    }
  },
  methods: {
    async mixPokemonArray () {
      this.pokemonArr = await getPokemonOptions()
      const randomInt = Math.floor(Math.random() * (this.pokemonArr.length))
      this.pokemon = this.pokemonArr[randomInt]
    },
    checkAnswer (selectedId) {
      this.showPokemon = true
      this.showAnswers = true
      if (selectedId === this.pokemon.id) {
        this.message = `Correct! ${this.pokemon.name}`
      } else {
        this.message = `Opps... it was ${this.pokemon.name}`
      }
    },
    newGame () {
      this.showPokemon = false
      this.showAnswers = false
      this.pokemonArr = []
      this.pokemon = null
      this.mixPokemonArray()
    }
  },
  mounted () {
    this.mixPokemonArray()
  }
}
</script>

<template>
  <main>
  <h1 v-if='!pokemon'>Loading...</h1>
  <div v-else>
    <h1>Who's that Pokemon?</h1>
    <PokemonImage  :pokemonId='pokemon.id' :showPokemon='showPokemon'/>
    <PokemonOptions :pokemons='pokemonArr' @selection='checkAnswer' />
    <div v-if='showAnswers' class='answer-container'>
      <h2 class='fade-in'>{{message}}</h2>
      <button @click='newGame'>New Game</button>
    </div>
  </div>
  </main>
</template>

<style scoped>
  main{
    display:flex;
    flex-direction: column;
    align-items: center;
  }
  .answer-container{
    display:flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;
  }
</style>
