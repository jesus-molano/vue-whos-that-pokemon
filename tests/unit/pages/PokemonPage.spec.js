import PokemonPage from '@/pages/PokemonPage'
import { shallowMount } from '@vue/test-utils'
import { pokemons } from '../mocks/pokemons.mock'

describe('PokemonPage component', () => {
  let wrapper
  let mixPokemonArraySpy
  beforeEach(() => {
    wrapper = shallowMount(PokemonPage)
    mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
  })
  test('should match with the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
  test('should call mixPokemonArray on mounted', () => {
    expect(mixPokemonArraySpy).toHaveBeenCalled()
  })
  test('should match with the snapshot when pokemons is loaded', () => {
    const wrapper = shallowMount(PokemonPage, {
      data () {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
  test('should show the components PokemonImage and PokemonOptions', () => {
    const wrapper = shallowMount(PokemonPage, {
      data () {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })
    const image = wrapper.find('pokemon-image-stub')
    const options = wrapper.find('pokemon-options-stub')
    expect(image.exists()).toBeTruthy()
    expect(options.exists()).toBeTruthy()
    expect(image.attributes('pokemonid')).toBe('1')
    expect(options.attributes('pokemons')).toBeTruthy()
  })
  test('should check the answer', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data () {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: ''
        }
      }
    })
    await wrapper.vm.checkAnswer(1)
    expect(wrapper.find('h2').exists()).toBeTruthy()
    expect(wrapper.vm.showPokemon).toBeTruthy()
    expect(wrapper.vm.message).toBe(`Correct! ${pokemons[0].name}`)
    await wrapper.vm.checkAnswer(10)
    expect(wrapper.vm.message).toBe(`Opps... it was ${pokemons[0].name}`)
  })
})
