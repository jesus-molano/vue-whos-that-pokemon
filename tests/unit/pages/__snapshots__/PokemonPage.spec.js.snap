// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component should match with the snapshot 1`] = `
<main>
  <h1>Loading...</h1>
</main>
`;

exports[`PokemonPage component should match with the snapshot when pokemons is loaded 1`] = `
<main>
  <div>
    <h1>Who's that Pokemon?</h1>
    <pokemon-image-stub pokemonid="1" showpokemon="false"></pokemon-image-stub>
    <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
    <!--v-if-->
  </div>
</main>
`;
